### PR TITLE
avoid allocation of SafeFreeSslCredentials and SafeDeleteSslContext on Linux

### PR DESCRIPTION
**Pull request by wfurt**
`SafeFreeSslCredentials` really have no meaning on Linux (unlike Schannel) and we simple drag copy of some properties from `sslAuthenticationOptions`.  The only difference is that we carte copy of certificates handle and key early so it would probably work even if disposed. I moved existing code to `AllocateSslContext` and we will get the handles when we carte the TLS session. 

`SafeDeleteSslContext` is basically wrapper that only stores reference to another `SafeHandle`. To avoid that, I made `SafeSslHandle` subclass from `SafeDeleteSslContext` so I can use single object. There may be better way to do it. For one, I was thinking about actually merging but the benefits seems same and it would be much bigger change.

### CODE REVIEW

File: src/libraries/System.Net.Quic/src/System.Net.Quic.csproj [hunk 1]
```diff
@@ -90,6 +90,12 @@
     <Compile Include="$(CommonPath)Microsoft\Win32\SafeHandles\Asn1SafeHandles.Unix.cs" Link="Common\Microsoft\Win32\SafeHandles\Asn1SafeHandles.Unix.cs" />
     <Compile Include="$(CommonPath)Microsoft\Win32\SafeHandles\SafeHandleCache.cs" Link="Common\Microsoft\Win32\SafeHandles\SafeHandleCache.cs" />
     <Compile Include="$(CommonPath)System\Net\SocketAddressPal.Unix.cs" Link="Common\System\Net\SocketAddressPal.Unix.cs" />
+    <Compile Include="$(CommonPath)System\Net\Security\Unix\SafeDeleteSslContext.cs" Link="Common\System\Net\Security\Unix\SafeDeleteSslContext.cs" />
```
Comment (ManickaP) [inline src/libraries/System.Net.Quic/src/System.Net.Quic.csproj:93]: :+1:

File: src/libraries/Common/src/Interop/Unix/System.Security.Cryptography.Native/Interop.OpenSsl.cs [hunk 2]
```diff
@@ -821,6 +814,42 @@ private static void BioWrite(SafeBioHandle bio, ReadOnlySpan<byte> buffer)
...
+        private static void SetSslCertificate(SafeSslContextHandle contextPtr, X509Certificate2 cert)
+        {
+            SafeEvpPKeyHandle? certKeyHandle = null;
+
+            using (RSAOpenSsl? rsa = (RSAOpenSsl?)cert.GetRSAPrivateKey())
+            {
+                if (rsa != null)
+                {
+                    certKeyHandle = rsa.DuplicateKeyHandle();
+                    Interop.Crypto.CheckValidOpenSslHandle(certKeyHandle);
```
Comment (stephentoub) [inline src/libraries/Common/src/Interop/Unix/System.Security.Cryptography.Native/Interop.OpenSsl.cs:826]: If this throws because certKeyHandle.IsInvalid is true, should certKeyHandle be disposed?

File: src/libraries/Common/src/Interop/Unix/System.Security.Cryptography.Native/Interop.OpenSsl.cs [hunk 3]
```diff
@@ -821,6 +814,42 @@ private static void BioWrite(SafeBioHandle bio, ReadOnlySpan<byte> buffer)
...
+            }
+
+            if (certKeyHandle == null)
+            {
+                using (ECDsaOpenSsl? ecdsa = (ECDsaOpenSsl?)cert.GetECDsaPrivateKey())
+                {
+                    if (ecdsa != null)
+                    {
+                        certKeyHandle = ecdsa.DuplicateKeyHandle();
+                        Interop.Crypto.CheckValidOpenSslHandle(certKeyHandle);
```
Comment (stephentoub) [inline src/libraries/Common/src/Interop/Unix/System.Security.Cryptography.Native/Interop.OpenSsl.cs:837]: Ditto

File: src/libraries/Common/src/System/Net/Security/Unix/SafeDeleteContext.cs [hunk 4]
```diff
@@ -29,15 +29,24 @@ protected SafeDeleteContext(SafeFreeCredentials credential)
...
         }
 
+        public SafeDeleteContext(IntPtr handle) : base(handle, true)
+        {
+            _credential = null;
+        }
+
+        public SafeDeleteContext(IntPtr handle, bool ownsHandle) : base(handle, ownsHandle)
+        {
+            _credential = null;
```
Comment (stephentoub) [inline src/libraries/Common/src/System/Net/Security/Unix/SafeDeleteContext.cs:39]: Are either of these `_credential = null` assignments necessary?  The default value will be null.

File: src/libraries/System.Net.Security/src/System/Net/Security/SslStream.Protocol.cs [hunk 5]
```diff
@@ -835,6 +835,8 @@ private SecurityStatusPal GenerateToken(ReadOnlySpan<byte> inputBuffer, ref byte
 
         internal SecurityStatusPal Renegotiate(out byte[]? output)
         {
+            ArgumentNullException.ThrowIfNull(_securityContext);
```
Comment (stephentoub) [inline src/libraries/System.Net.Security/src/System/Net/Security/SslStream.Protocol.cs:838]: _securityContext isn't an argument.  In what situations will we get here and find _securityContext is null?

File: src/libraries/System.Net.Security/src/System/Net/Security/SslStreamPal.Unix.cs [hunk 6]
```diff
@@ -150,25 +149,23 @@ public static void QueryContextStreamSizes(SafeDeleteContext? securityContext, o
...
         {
-            Debug.Assert(!credential.IsInvalid);
-
             byte[]? output = null;
             int outputSize = 0;
 
             try
             {
                 if ((null == context) || context.IsInvalid)
                 {
```
Comment (stephentoub) [inline src/libraries/System.Net.Security/src/System/Net/Security/SslStreamPal.Unix.cs:164]: If context.IsInvalid, should we Dispose it?

File: src/libraries/Common/src/Interop/Unix/System.Security.Cryptography.Native/Interop.OpenSsl.cs (hunk 2 above)
Comment (wfurt) [reply src/libraries/Common/src/Interop/Unix/System.Security.Cryptography.Native/Interop.OpenSsl.cs:826]: good questions. This is essentially copy from 
https://github.com/dotnet/runtime/blob/9af8d302389305d2efaf1fd8db5c1dea0a246900/src/libraries/Common/src/System/Net/Security/Unix/SafeFreeSslCredentials.cs#L46-L93

Since this all comes from single object e.g. `X509Certificate2` I'm not sure if that is possible. 
cc: @bartonjs  

I'de be happy to add code to deal with it instead of depending on finalizer.

Comment (wfurt) [reply src/libraries/Common/src/Interop/Unix/System.Security.Cryptography.Native/Interop.OpenSsl.cs:826]: good questions. This is essentially copy from 
https://github.com/dotnet/runtime/blob/9af8d302389305d2efaf1fd8db5c1dea0a246900/src/libraries/Common/src/System/Net/Security/Unix/SafeFreeSslCredentials.cs#L46-L93

Since this all comes from single object e.g. `X509Certificate2` I'm not sure if that is possible. 
cc: @bartonjs  

I'de be happy to add code to deal with it instead of depending on finalizer.

File: src/libraries/System.Net.Security/src/System/Net/Security/SslStream.Protocol.cs (hunk 5 above)
Comment (wfurt) [reply src/libraries/System.Net.Security/src/System/Net/Security/SslStream.Protocol.cs:838]: It would happen if this is called before `AuthenticateAs*`. However, `NegotiateClientCertificateAsync` has guard `ThrowIfExceptionalOrNotAuthenticated`.  I can remove it or turn it into `Debug.Assert`

File: src/libraries/Common/src/Interop/Unix/System.Security.Cryptography.Native/Interop.OpenSsl.cs (hunk 2 above)
Comment (bartonjs) [reply src/libraries/Common/src/Interop/Unix/System.Security.Cryptography.Native/Interop.OpenSsl.cs:826]: The code has changed a bit in the years since.  RSAOpenSsl::DuplicateKeyHandle will never return an IsInvalid SafeHandle, it would have already disposed the object and thrown.

https://github.com/dotnet/runtime/blob/eef6c7f6cb7f73163313d349fbb796f37b4c3996/src/libraries/Common/src/Interop/Unix/System.Security.Cryptography.Native/Interop.EvpPkey.cs#L21-L38

Comment (wfurt) [reply src/libraries/Common/src/Interop/Unix/System.Security.Cryptography.Native/Interop.OpenSsl.cs]: I realized we don't need to do this for each session. Moved the handle logic to `SslStreamCertificateContext.Linux` so it is done once upfront.

File: src/libraries/System.Net.Security/src/System/Net/Security/SslStream.Protocol.cs (hunk 5 above)
Comment (wfurt) [reply src/libraries/System.Net.Security/src/System/Net/Security/SslStream.Protocol.cs]: updated to Debug.Assert

File: src/libraries/Common/src/System/Net/Security/Unix/SafeDeleteContext.cs (hunk 4 above)
Comment (wfurt) [reply src/libraries/Common/src/System/Net/Security/Unix/SafeDeleteContext.cs]: no. I did deeper cut and removed `_credential`  completely so we don't drag unneeded property.

File: src/libraries/System.Net.Security/src/System/Net/Security/SslStreamPal.Unix.cs (hunk 6 above)
Comment (wfurt) [reply src/libraries/System.Net.Security/src/System/Net/Security/SslStreamPal.Unix.cs:164]: I look through the code and I think the only one way how this could happen after creation is Disposing the context.
I think we added some of the check while back when we had issue with using closed handles.

**Comment by wfurt** [comment]
This is ready for another pass @stephentoub.

File: src/libraries/Common/src/System/Net/Security/Unix/SafeDeleteNegoContext.cs [hunk 7]
```diff
@@ -59,6 +63,9 @@ public SafeDeleteNegoContext(SafeFreeNegoCredentials credential, string targetNa
                 Dispose();
                 throw;
             }
+            _credential = credential;
+            bool ignore = false;
+            _credential.DangerousAddRef(ref ignore);
```
Comment (stephentoub) [inline src/libraries/Common/src/System/Net/Security/Unix/SafeDeleteNegoContext.cs:68]: Where is DangerousRelease called?

Comment (wfurt) [reply src/libraries/Common/src/System/Net/Security/Unix/SafeDeleteNegoContext.cs:70]: I used `_credential.Dispose` but that was probably incorrect. Since we don't own the credentials here, we should not Dispose. I modified the code to drop the reference count instead.

Comment (stephentoub) [reply src/libraries/Common/src/System/Net/Security/Unix/SafeDeleteNegoContext.cs:70]: Dispose and DangerousRelease are different.  In fact, if there's an outstanding reference count on the instance, Dispose will be a nop until that ref count drops to 0.

Comment (wfurt) [reply src/libraries/Common/src/System/Net/Security/Unix/SafeDeleteNegoContext.cs:70]: yes, I realized that after the comment. I was lured by existing code in Dispose without realizing the pattern was different ;(

File: src/libraries/Common/src/System/Net/Security/Unix/SafeDeleteNegoContext.cs [hunk 8]
```diff
@@ -90,6 +102,12 @@ protected override void Dispose(bool disposing)
                     _acceptorCredential.Dispose();
                     _acceptorCredential = null;
                 }
+
+                if (_credential != null)
+                {
+                    _credential.DangerousRelease();
```
Comment (stephentoub) [inline src/libraries/Common/src/System/Net/Security/Unix/SafeDeleteNegoContext.cs:108]: This needs to be done even if disposing is false.  Otherwise, _credential will be forever leaked.

File: src/libraries/Common/src/System/Net/Security/Unix/SafeDeleteNegoContext.cs [hunk 9]
```diff
@@ -42,14 +43,17 @@ public SafeGssContextHandle GssContext
         }
 
         public SafeDeleteNegoContext(SafeFreeNegoCredentials credential)
-            : base(credential)
+            : base(IntPtr.Zero)
         {
             Debug.Assert((null != credential), "Null credential in SafeDeleteNegoContext");
+            _credential = credential;
```
Comment (stephentoub) [inline src/libraries/Common/src/System/Net/Security/Unix/SafeDeleteNegoContext.cs:49]: This should be done only after the call to DangerousAddRef.  Otherwise, if the DangerousAddRef throws, the finalizer will still try to call DangerousRelease on the _credential.

File: src/libraries/System.Net.Security/src/System/Net/Security/SslStreamCertificateContext.Linux.cs [hunk 10]
```diff
@@ -32,11 +35,44 @@ private SslStreamCertificateContext(X509Certificate2 target, X509Certificate2[]
...
+
+            using (RSAOpenSsl? rsa = (RSAOpenSsl?)target.GetRSAPrivateKey())
+            {
+                if (rsa != null)
+                {
+                    KeyHandle = rsa.DuplicateKeyHandle();
+                }
+            }
+
+            if (KeyHandle== null)
```
Comment (stephentoub) [inline src/libraries/System.Net.Security/src/System/Net/Security/SslStreamCertificateContext.Linux.cs:48]:
```suggestion
            if (KeyHandle == null)
```

File: src/libraries/System.Net.Security/src/System/Net/Security/SslStreamCertificateContext.Linux.cs [hunk 11]
```diff
@@ -32,11 +35,44 @@ private SslStreamCertificateContext(X509Certificate2 target, X509Certificate2[]
...
+                    {
+                        KeyHandle = ecdsa.DuplicateKeyHandle();
+                    }
+                }
+            }
+
+            if (KeyHandle== null)
+            {
+                throw new NotSupportedException(SR.net_ssl_io_no_server_cert);
+            }
```
Comment (stephentoub) [inline src/libraries/System.Net.Security/src/System/Net/Security/SslStreamCertificateContext.Linux.cs:62]:
```suggestion
                    if (ecdsa != null)
                    {
                        KeyHandle = ecdsa.DuplicateKeyHandle();
                        if (KeyHandle == null)
                        {
                            throw new NotSupportedException(SR.net_ssl_io_no_server_cert);
                        }
                    }
                }
            }


```

File: src/libraries/System.Net.Security/src/System/Net/Security/SslStreamCertificateContext.Linux.cs [hunk 12]
```diff
@@ -32,11 +35,44 @@ private SslStreamCertificateContext(X509Certificate2 target, X509Certificate2[]
...
+                using (ECDsaOpenSsl? ecdsa = (ECDsaOpenSsl?)target.GetECDsaPrivateKey())
+                {
+                    if (ecdsa != null)
+                    {
+                        KeyHandle = ecdsa.DuplicateKeyHandle();
+                    }
+                }
+            }
+
+            if (KeyHandle== null)
```
Comment (stephentoub) [inline src/libraries/System.Net.Security/src/System/Net/Security/SslStreamCertificateContext.Linux.cs:59]:
```suggestion
            if (KeyHandle == null)
```

File: src/libraries/System.Net.Security/src/System/Net/Security/SslStreamCertificateContext.Linux.cs (hunk 11 above)
Comment (stephentoub) [inline src/libraries/System.Net.Security/src/System/Net/Security/SslStreamCertificateContext.Linux.cs:62]:
```suggestion

                if (KeyHandle == null)
                {
                    throw new NotSupportedException(SR.net_ssl_io_no_server_cert);
                }
            }
```

File: src/libraries/System.Net.Security/src/System/Net/Security/SslStreamCertificateContext.Linux.cs [hunk 13]
```diff
@@ -32,11 +35,44 @@ private SslStreamCertificateContext(X509Certificate2 target, X509Certificate2[]
...
         }
 
         internal static SslStreamCertificateContext Create(X509Certificate2 target) =>
             Create(target, null, offline: false, trust: null, noOcspFetch: true);
 
+        ~SslStreamCertificateContext()
+        {
+            CertificateHandle?.Dispose();
+            KeyHandle?.Dispose();
+        }
```
Comment (stephentoub) [inline src/libraries/System.Net.Security/src/System/Net/Security/SslStreamCertificateContext.Linux.cs:74]: Why is this needed?  CertificateHandle and KeyHandle don't have their own finalizers?

**Review by stephentoub — COMMENTED** [review]
Mostly looks good, but some SafeHandle-related issues to be addressed before it can be merged.

File: src/libraries/Common/src/System/Net/Security/Unix/SafeDeleteNegoContext.cs (hunk 9 above)
Comment (wfurt) [reply src/libraries/Common/src/System/Net/Security/Unix/SafeDeleteNegoContext.cs:55]: make sense. updated.

File: src/libraries/System.Net.Security/src/System/Net/Security/SslStreamCertificateContext.Linux.cs (hunk 13 above)
Comment (wfurt) [reply src/libraries/System.Net.Security/src/System/Net/Security/SslStreamCertificateContext.Linux.cs]: no specific reason. removed.
I was experimenting with the ideas since this is not Disposable. Leaving it to GC is probably best.

**Comment by wfurt** [comment]
I made updates to address comments @stephentoub. Can you please take another look?

File: src/libraries/Common/src/System/Net/Security/Unix/SafeDeleteNegoContext.cs [hunk 14]
```diff
@@ -42,14 +43,21 @@ public SafeGssContextHandle GssContext
...
-            : base(credential)
+            : base(IntPtr.Zero)
         {
             Debug.Assert((null != credential), "Null credential in SafeDeleteNegoContext");
+            bool added = false;
+            credential.DangerousAddRef(ref added);
+            if (!added)
+            {
+                throw new ObjectDisposedException(nameof(SafeFreeNegoCredentials));
+            }
```
Comment (stephentoub) [inline src/libraries/Common/src/System/Net/Security/Unix/SafeDeleteNegoContext.cs:54]: `added` will never be false.  If the operation can't add, it throws an exception.
```suggestion
```

File: src/libraries/System.Net.Security/src/System/Net/Security/SslStreamCertificateContext.Linux.cs [hunk 15]
```diff
@@ -32,6 +35,33 @@ private SslStreamCertificateContext(X509Certificate2 target, X509Certificate2[]
             IntermediateCertificates = intermediates;
             Trust = trust;
             SslContexts = new ConcurrentDictionary<SslProtocols, SafeSslContextHandle>();
+
+
```
Comment (stephentoub) [inline src/libraries/System.Net.Security/src/System/Net/Security/SslStreamCertificateContext.Linux.cs:39]:
```suggestion
```

**Review by stephentoub — APPROVED** [review]
